src/Adapter/FilesystemItems.php

```diff
@@ -0,0 +1,209 @@
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            return;
+        }
+
+        if (null === $version) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        try {
+            $highest = $this->get($id)->getVersion();
+        } catch (ItemNotFound $e) {
+            return;
+        }
+
+        if ($version->toInt() < $highest->toInt()) {
+            throw new UnexpectedVersionNumber($id, $version, $highest);
+        }
+
+        if (1 === $version->toInt()) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        $this->filesystem->remove("{$this->path}/{$id}/{$version->toInt()}.xml");
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $version) {
+            $versions = iterator_to_array($this->getVersions($id));
+            rsort($versions);
+            $version = $versions[0];
+        }
+
+        if (!$content = @fopen($file = "{$this->path}/{$id}/{$version->toInt()}.xml", 'rb')) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        if (!$hash = md5_file($file)) {
+            throw new RuntimeException("Failed to hash {$file}");
+        }
+
+        return new ItemVersion($id, $version, $content, $hash);
+    }
+
+    public function list(int $limit = 10, ?ItemId $startAt = null) : ItemListPage
+    {
+        $ids = $this->getIds();
+
+        if (null === $startAt) {
+            $ids = new LimitIterator($ids, 0, $limit + 1);
+        } else {
+            $found = null;
+            $i = 0;
+            foreach ($ids as $id) {
+                if ($id == $startAt) {
+                    $found = $i;
+                    break;
+                }
+                $i++;
+            }
+
+            if (null === $found) {
+                $ids = new EmptyIterator();
+            } else {
+                $ids = new LimitIterator($ids, $found, $limit + 1);
+            }
+        }
+
+        $ids = array_values(iterator_to_array($ids));
+
+        if (count($ids) > $limit) {
+            $nextId = array_pop($ids);
+        }
+
+        return new ItemListPage($ids, $nextId ?? null);
+    }
+
+    public function count() : int
+    {
+        return iterator_count($this->getIds());
+    }
+
+    public function getIterator() : Traversable
+    {
+        $items = new MapIterator(
+            $this->getIds(),
+            function (ItemId $id) : ItemVersion {
+                return $this->get($id);
+            }
+        );
+
+        foreach ($items as $item) {
+            yield $item;
+        }
+    }
+
+    private function getIds() : Iterator
+    {
+        return new MapIterator(
+            new CallbackFilterIterator(
+                new FilesystemIterator("{$this->path}/"),
```

Comment:
unclear what the order is here, is it deterministic?

Reply:
Ordering is an interesting question. The API spec doesn't currently define an order. So, the order returned currently doesn't matter, but has to be deterministic for the paging to work. This is currently true, but the question is what to happen to new items while you're paging. Should you expect to see them? In this implementation, it's ordered alphabetically on the ID, so you may seem them depending on what it is.

No defined order makes the listing API not very useful, expect for mass imports elsewhere (eg search reindexing).

Reply:
:+1: , would just be helpful to document the ordering somewhere here, or to make it explicit if it's configurable